Feature task1

ilushaTigach opened the pull request.

ilushaTigach committed:
initial commit
Сделал: CRUD, миграции Liquibase

ilushaTigach committed:
Delete .idea directory

ilushaTigach committed:
Delete Dockerfile

GUR-ok commented:
Почему мр на вливание в мейн?

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/models/ProductEntity.java:
```diff
@@ -0,0 +1,57 @@
+    private Integer quantity;
+
+    @Column(name = "last_modified")
+    private LocalDateTime lastModified;
```
offsetDateTime

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/models/ProductEntity.java:
```diff
@@ -0,0 +1,57 @@
+
+@Entity
+@Table(name = "products")
+@Data
```
toString переопределить вручную

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/models/ProductEntity.java:
```diff
@@ -0,0 +1,57 @@
+
+    @Min(value = 1, message = "Quantity cannot be negative")
+    @Column(name = "quantity")
+    private Integer quantity;
```
BigDecimal

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/models/ProductEntity.java:
```diff
@@ -0,0 +1,57 @@
+
+    @Min(value = 1, message = "Price cannot be negative")
+    @Column(name = "price")
+    private Integer price;
```
BigDecimal

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/models/ProductEntity.java:
```diff
@@ -0,0 +1,57 @@
+    private String description;
+
+    @NotNull(message = "ProductEntity type cannot be null")
+    @Column(name = "productType")
```
@Enumerated(EnumType.STRING)

GUR-ok replied:
, nullable = false)

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/models/ProductResponse.java:
```diff
@@ -0,0 +1,38 @@
+    @NotBlank(message = "Title cannot be blank")
+    private String title;
+
+    @NotBlank(message = "Article cannot be blank")
```
валидацию на респонс не ставим

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/services/ProductService.java:
```diff
@@ -0,0 +1,87 @@
+        String article = productDto.getArticle();
+        if (article != null) {
+            ProductEntity existingProduct = productRepository.findByArticle(article);
+            if (existingProduct != null) {
```
через optional обработать

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/services/ProductService.java:
```diff
@@ -0,0 +1,87 @@
+        if (article != null) {
+            ProductEntity existingProduct = productRepository.findByArticle(article);
+            if (existingProduct != null) {
+                throw new IllegalArgumentException("Product with article " + article + " already exists.");
```
кастомное исключение

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/services/ProductService.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    @Transactional
+    public ProductDto saveProduct(ProductDto productDto) {
```
давай вернем просто id

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/services/ProductService.java:
```diff
@@ -0,0 +1,87 @@
+
+    public void deleteProductById(UUID id) {
+        productRepository.findById(id).orElseThrow(() -> new NotFoundProductException("id", id.toString()));
+        productRepository.deleteById(id);
```
лишний запрос, тк deleteById Итак делает select. Включить логирование sql запросов hibernate и проконтролировать число запросов

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/services/ProductService.java:
```diff
@@ -0,0 +1,87 @@
+        if (article != null && !article.equals(productEntity.getArticle())) {
+            ProductEntity existingProduct = productRepository.findByArticle(article);
+            if (existingProduct != null && !existingProduct.getId().equals(id)) {
+                throw new IllegalArgumentException("Product with article " + article + " already exists.");
```
в сообщении вернуть Id дубля

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/services/ProductService.java:
```diff
@@ -0,0 +1,87 @@
+        productEntity.setProductType(updateProductDto.getProductType());
+        productEntity.setPrice(updateProductDto.getPrice());
+        productEntity.setQuantity(updateProductDto.getQuantity());
+        return updateProductMapper.toDto(productRepository.save(productEntity));
```
вернуть void или id

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/services/ProductService.java:
```diff
@@ -0,0 +1,87 @@
+    private final UpdateProductMapper updateProductMapper;
+
+    public List<ProductDto> findAllProducts() {
+        Pageable pageable = PageRequest.of(0, 3);
```
pageable принимать целиком в контроллере, тут не создавать

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/repositories/ProductRepository.java:
```diff
@@ -0,0 +1,17 @@
+@Repository
+public interface ProductRepository extends JpaRepository<ProductEntity, UUID> {
+
+    Page<ProductEntity> findAll(Pageable pageable);
```
должен быть доступен из коробки

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/resources/ProductResource.java:
```diff
@@ -0,0 +1,35 @@
+
+import io.swagger.v3.oas.annotations.Operation;
+import io.swagger.v3.oas.annotations.tags.Tag;
+import org.springframework.web.bind.annotation.*;
```
импорты везде развернуть и настроить идею, чтоб не схлопывала

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/repositories/ProductRepository.java:
```diff
@@ -0,0 +1,17 @@
+package org.telatenko.storagesevicedomain.repositories;
```
сделаем пакет persistence
в нем entity и repository

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/controllers/ProductController.java:
```diff
@@ -0,0 +1,55 @@
+package org.telatenko.storagesevicedomain.controllers;
```
название пакетов в ед числе

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/dto/ProductType.java:
```diff
@@ -0,0 +1,7 @@
+
+public enum ProductType {
+
+    FRUIT, TECH, DRINK, FURNITURE;
```
енумчик в отдельный пакет enum

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/handler/ExceptionHandlerController.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.Map;
+
+@RestControllerAdvice
+public class ExceptionHandlerController {
```
в пакет exception

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/services/ProductService.java:
```diff
@@ -0,0 +1,87 @@
+@Service
+@RequiredArgsConstructor
+@Validated
+public class ProductService {
```
создать интерфейс и реализацию

GUR-ok commented on storage-service-domain/src/main/java/org/telatenko/storagesevicedomain/services/ProductService.java:
```diff
@@ -0,0 +1,87 @@
+        return findAllProductsMapper.toDto(productRepository.findAll(pageable).getContent());
+    }
+
+    public ProductDto findProductById(UUID id) {
```
Transactional(readOnly

GUR-ok commented on storage-service-domain/src/main/resources/application-local.yml:
```diff
@@ -0,0 +1,14 @@
+    username: root
+    password: ${PWD}
+    driver-class-name: org.postgresql.Driver
+  jpa:
```
jpa:
    hibernate:
      ddl-auto: validate

GUR-ok commented on storage-service-domain/docker-compose.yml:
```diff
@@ -0,0 +1,46 @@
+version: "3.9"
+
```
простой докер файл тоже добавить

ilushaTigach committed:
commit №2
-Миграции переписаны в sql и xml
-application-local теперь рабоатет
-Сделал Dockerfile и docker-compose.yml
-Тесты пока что не писал
-Не понял как сделать Пагинацию праввильно, чтобы из коробки доставлась

ilushaTigach committed:
Merge remote-tracking branch 'origin/feature_task1' into feature_task1

# Conflicts:
#	.idea/compiler.xml
#	.idea/encodings.xml
#	.idea/misc.xml
#	.idea/sqldialects.xml
#	.idea/workspace.xml
#	Dockerfile

ilushaTigach committed:
commit №2
-Миграции переписаны в sql и xml
-application-local теперь рабоатет
-Сделал Dockerfile и docker-compose.yml
-Тесты пока что не писал
-Не понял как сделать Пагинацию праввильно, чтобы из коробки доставлась

ilushaTigach committed:
Delete .idea directory

ilushaTigach committed:
commit №3
-Написал тесты
-написал javadoc
-добавил коллекцию из Postman

ilushaTigach committed:
Merge remote-tracking branch 'origin/feature_task1' into feature_task1

ilushaTigach committed:
commit №4
-Написал 2 шедулера
-Написал 2 аннотации на измерение времени

GUR-ok commented on src/main/java/org/telatenko/storagesevicedomain/scheduler/OptimizedPriceIncreaseScheduler.java:
```diff
@@ -0,0 +1,104 @@
+                    products.add(product);
+                }
+            }
+            System.out.println("Закончили селекты");
```
sout Не должно быть. используй логирование если нужно

GUR-ok commented on src/main/java/org/telatenko/storagesevicedomain/scheduler/OptimizedPriceIncreaseScheduler.java:
```diff
@@ -0,0 +1,104 @@
+
+            connection.commit();
+        } catch (SQLException e) {
+            throw new RuntimeException("Failed to update prices", e);
```
Rollback Не нужно делать? 
А если упадет c npe или illegalArgumentException или castException например на вызове .getInt(1), то connection закроется?

GUR-ok commented on src/main/java/org/telatenko/storagesevicedomain/scheduler/OptimizedPriceIncreaseScheduler.java:
```diff
@@ -0,0 +1,104 @@
+    private JdbcTemplate jdbcTemplate;
+
+    @Value("${app.scheduling.priceMultiplier}")
+    private String priceMultiplier;
```
почему не сразу BigDecimal?

GUR-ok commented on src/main/java/org/telatenko/storagesevicedomain/scheduler/OptimizedPriceIncreaseScheduler.java:
```diff
@@ -0,0 +1,104 @@
+        String selectSql = "SELECT id, price FROM products ORDER BY id LIMIT ? OFFSET ?";
+        String updateSql = "UPDATE products SET price = ? WHERE id = ?";
+
+        try (PreparedStatement selectStmt = connection.prepareStatement(selectSql);
```
PreparedStatement нужно создать 1 раз, а не столько раз, сколько у тебя бачей. этот метод updateBatch вызывается неск. раз. соответственно и стейтмент создается неск.раз.

GUR-ok commented on src/main/java/org/telatenko/storagesevicedomain/scheduler/OptimizedPriceIncreaseScheduler.java:
```diff
@@ -0,0 +1,104 @@
+                    ProductEntity product = new ProductEntity();
+                    product.setId(UUID.fromString(rs.getString("id")));
+                    product.setPrice(rs.getBigDecimal("price"));
+                    products.add(product);
```
не нужно сохранять объекты в массив. тебе вообще не нужны тут ProductEntity. Ты можешь сразу в файл писать построчно

GUR-ok commented on src/main/java/org/telatenko/storagesevicedomain/scheduler/OptimizedPriceIncreaseScheduler.java:
```diff
@@ -0,0 +1,104 @@
+            updateStmt.executeBatch();
+
+        } catch (SQLException | IOException e) {
+            try {
```
catch в catch не нужно делать.
Почему нельзя так?
```
                    } catch (Exception e) {
                        connection.rollback();
                        throw new RuntimeException(e);
                    }
```

GUR-ok commented on src/main/java/org/telatenko/storagesevicedomain/scheduler/OptimizedPriceIncreaseScheduler.java:
```diff
@@ -0,0 +1,104 @@
+                updateStmt.setBigDecimal(1, newPrice);
+                updateStmt.setObject(2, product.getId());
+                updateStmt.addBatch();
+                System.out.println(String.format("Product ID: %s, Old Price: %s, New Price: %s\n",
```
sout убираем

GUR-ok commented on src/main/java/org/telatenko/storagesevicedomain/scheduler/OptimizedPriceIncreaseScheduler.java:
```diff
@@ -0,0 +1,104 @@
+                }
+            }
+            System.out.println("Закончили селекты");
+            for (ProductEntity product : products) {
```
этого цикла не будет. пишем в файл сразу в цикле while (rs.next())

GUR-ok commented on src/main/java/org/telatenko/storagesevicedomain/scheduler/OptimizedPriceIncreaseScheduler.java:
```diff
@@ -0,0 +1,104 @@
+
+        try (PreparedStatement selectStmt = connection.prepareStatement(selectSql);
+             PreparedStatement updateStmt = connection.prepareStatement(updateSql);
+             BufferedWriter writer = new BufferedWriter(new FileWriter(LOG_FILE_PATH, true))) {
```
BufferedWriter fileWriter тоже нужно в трай с ресурсами.

GUR-ok commented on src/main/java/org/telatenko/storagesevicedomain/scheduler/OptimizedPriceIncreaseScheduler.java:
```diff
@@ -0,0 +1,104 @@
+            ResultSet resultSet = connection.prepareStatement("SELECT COUNT(*) FROM products").executeQuery();
+            resultSet.next();
+            int totalProducts = resultSet.getInt(1);
+            for (int start = 0; start < totalProducts; start += BATCH_SIZE) {
```
ты уверен что последний бач экзекьютится? Ну например у нас 150 элементов, размер бача 100. Вот последние 50 отработают?

GUR-ok commented on src/main/java/org/telatenko/storagesevicedomain/persistence/ProductRepository.java:
```diff
@@ -0,0 +1,46 @@
+    List<ProductEntity> findAllForScheduler(LockModeType lockMode);
+
+    @Query(value = "SELECT p.* FROM (SELECT *, ROW_NUMBER() OVER (ORDER BY id) AS rn FROM products) p WHERE p.rn BETWEEN :startId AND :endId", nativeQuery = true)
+    List<ProductEntity> findAllInRange(@Param("startId") long startId, @Param("endId") long endId);
```
это где используется?

GUR-ok commented on src/main/java/org/telatenko/storagesevicedomain/exeption/NotFoundProductException.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Исключение, которое выбрасывается, когда продукт с указанным идентификатором не найден.
+ */
+public class NotFoundProductException extends RuntimeException {
```
ProductNotFoundException

GUR-ok commented on src/main/java/org/telatenko/storagesevicedomain/exeption/ExceptionHandlerController.java:
```diff
@@ -0,0 +1,101 @@
+ * Этот класс предоставляет методы для обработки различных исключений и возвращения соответствующих HTTP ответов.
+ */
+@RestControllerAdvice
+public class ExceptionHandlerController {
```
Это не контроллер ) Либо просто ExceptionHandler, либо ControllerExceptionHandler, либо GlobalExceptionHandler. В англ. языке слово, стоящее на последнем месте является определяющим. Т.е. например "цена товара" будет как ProductPrice, а не PriceProduct или еще хуже PriceOfProduct

GUR-ok commented on src/main/java/org/telatenko/storagesevicedomain/acpect/TransactionalExecutionTimeAspect.java:
```diff
@@ -0,0 +1,27 @@
+    @Around("@annotation(org.telatenko.storagesevicedomain.annotation.MeasureTransactionalExecutionTime) " +
+            "&& @annotation(org.springframework.transaction.annotation.Transactional)")
+    public Object measureExecutionTime(ProceedingJoinPoint joinPoint) throws Throwable {
+        long start = System.currentTimeMillis();
```
а в чем отличие от ExecutionTimeAspect ? Ты уверен что время замеряется в момент коммита транзакции, а не в момент окончания выполнения базового тела метода? проверь на обычном шедулере. У тебя в консоль выведется время, а в бд данные еще не обновятся (посмотри в бобре в этот момент)

ilushaTigach committed:
commit №5
Внёс исправления
